src/Rbac.php

```diff
@@ -67,65 +68,52 @@ public function addRole($child, $parents = null)
                 if ($this->createMissingRoles && ! $this->hasRole($parent)) {
                     $this->addRole($parent);
                 }
-                $this->getRole($parent)->addChild($child);
+                if (is_string($parent)) {
+                    $parent = $this->getRole($parent);
+                }
+                $parent->addChild($role);
             }
         }
-
-        $this->children[] = $child;
-
-        return $this;
+        $this->roles[$role->getName()] = $role;
     }
 
     /**
-     * Is a child with $name registered?
+     * Is a role registered?
      *
-     * @param  \Zend\Permissions\Rbac\RoleInterface|string $objectOrName
+     * @param  RoleInterface|string $role
      * @return bool
      */
-    public function hasRole($objectOrName)
+    public function hasRole($role): bool
     {
-        try {
-            $this->getRole($objectOrName);
-
-            return true;
-        } catch (Exception\InvalidArgumentException $e) {
-            return false;
+        if (! is_string($role) && ! $role instanceof RoleInterface) {
+            throw new Exception\InvalidArgumentException(
+                'Role must be a string or implement Zend\Permissions\Rbac\RoleInterface'
+            );
         }
+        if (is_string($role)) {
+            return isset($this->roles[$role]);
+        }
+        $roleName = $role->getName();
+        return isset($this->roles[$roleName]) &&
+               $this->roles[$roleName] === $role;
```

Comment:
Push the `&&` to the next line, and indent only one additional level (4 spaces).

Reply:
Done